### PR TITLE
GAP-2103 - Edit section title

### PR DESCRIPTION
**Pull request by john-tco**
## Description

Adds ui for an admin to change a section title

[GAP-2103](https://technologyprogramme.atlassian.net/jira/software/projects/GAP/boards/216?selectedIssue=GAP-2103)

Related pr: https://github.com/cabinetoffice/gap-find-admin-backend/pull/162

Screenshot 2024-01-31 at 11 42 19

Screenshot 2024-01-31 at 11 42 45

Summary of the changes and the related issue. List any dependencies that are required for this change:

## Type of change

Please check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [ ] Unit Test

- [ ] Integration Test (if applicable)

- [ ] End to End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

# Checklist:

- [ ] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [ ] I have performed a self-review of my code.
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation where applicable.
- [ ] I have ran cypress tests and they all pass locally.

[GAP-2103]: https://technologyprogramme.atlassian.net/browse/GAP-2103?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/edit-title.page.tsx [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      body,
+    });
+
+  return QuestionPageGetServerSideProps({
+    context,
+    fetchPageData,
+    handleRequest,
+    jwt: getSessionIdFromCookies(context.req),
+    onErrorMessage: 'Failed to edit title, please try again later.',
+    onSuccessRedirectHref: `/build-application/${context.params?.applicationId}/${context.params?.sectionId}/`,
```
Comment (dominicwest) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/edit-title.page.tsx:40]:
```suggestion
    onSuccessRedirectHref: `/build-application/${applicationId}/${sectionId}`,
```

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/edit-title.page.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    sessionId: string
+  ) =>
+    updateSectionTitle({
+      sessionId,
+      applicationId,
+      sectionId,
+      body,
+    });
+
+  return QuestionPageGetServerSideProps({
```
Comment (dominicwest) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/edit-title.page.tsx:34]: Love that you're using this <3

File: packages/admin/src/services/SectionService.ts [hunk 3]
```diff
@@ -51,4 +52,22 @@ const updateSectionStatus = (
...
+const updateSectionTitle = async ({
+  sessionId,
+  applicationId,
+  sectionId,
+  body: { sectionTitle },
+}: UpdateSectionTitleProps) => {
+  const config: AxiosConfig = {
+    ...axiosSessionConfig(sessionId),
+  };
+  config.headers['Content-Type'] = 'application/json';
```
Comment (dominicwest) [inline packages/admin/src/services/SectionService.ts:64]: This really needed?

File: packages/admin/src/services/SectionService.ts [hunk 4]
```diff
@@ -51,4 +52,22 @@ const updateSectionStatus = (
...
+
+  await axios.patch(
+    `${BASE_APPLICATION_URL}/${applicationId}/sections/${sectionId}/title`,
+    { sectionTitle },
+    config
+  );
```
Comment (dominicwest) [inline packages/admin/src/services/SectionService.ts:70]:
```suggestion
    await axios.patch(
    `${BASE_APPLICATION_URL}/${applicationId}/sections/${sectionId}/title`,
    { sectionTitle },
    axiosSessionConfig(sessionId)
  );
```

**Review by dominicwest — CHANGES_REQUESTED** [review]

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/edit-title.page.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      <div className="govuk-!-padding-top-7">
+        <FlexibleQuestionPageLayout
+          formAction={formAction}
+          csrfToken={csrfToken}
+          fieldErrors={fieldErrors}
+        >
+          <h1 className="govuk-heading-l">Change the title of this section</h1>
+
+          <TextInput
+            defaultValue={sectionTitle}
```
Comment (dominicwest) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/edit-title.page.tsx:71]: Need to make use of the `previousValues` prop to retain previously entered input upon a validation error

File: packages/admin/src/services/SectionService.ts (hunk 4 above)
Comment (dominicwest) [reply packages/admin/src/services/SectionService.ts:70]: Tested locally - works for me

**Review by dominicwest — CHANGES_REQUESTED** [review]

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/edit-title.page.tsx (hunk 5 above)
Comment (john-tco) [reply packages/admin/src/pages/build-application/[applicationId]/[sectionId]/edit-title.page.tsx]: Good spot :)

**Review by dominicwest — APPROVED** [review]

**Pull request merged**